### PR TITLE
Proper arch start

### PR DESCRIPTION
**Pull request by alexurumov**
Attempting to merge proper-arch-start branch to main develop branch, as all of the requirements discussed have been implemented

### CODE REVIEW

File: src/services/user.service.ts [hunk 1]
```diff
@@ -1,51 +1,83 @@
...
+        return users.map(toUserDto);
+    }
+
+    async update (username: string, userDto: UserUpdateDto): Promise<UserDto> {
+        const [existing, all] = await Promise.all([this.userRepository.findByUsername(username), this.userRepository.findAll()]);
+
+        // Validate specific meeting requirements
+        validateUpdateUser(existing, userDto, all);
+
+        const id = existing!._id.toString();
```
Comment (petrovdimitar92) [inline src/services/user.service.ts]: `
let id;
if(existing) {
  id = existing._id.toString();
}
`

File: .eslintrc.json [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+        ]
+    },
+    "rules": {
+        "@typescript-eslint/indent": ["error", 4],
+        "@typescript-eslint/semi": ["error", "always"],
+        "@typescript-eslint/strict-boolean-expressions": "off",
+        "@typescript-eslint/no-invalid-void-type": "off",
+        "@typescript-eslint/ban-ts-comment": "off",
+        "@typescript-eslint/no-misused-promises": "off",
+        "@typescript-eslint/no-non-null-assertion": "off",
```
Comment (petrovdimitar92) [inline .eslintrc.json]: remove 'no-non-null-assertion'

File: src/entities/meeting.entity.ts [hunk 3]
```diff
@@ -1,19 +1,37 @@
...
+    Daily = 'daily',
+    Weekly = 'weekly',
+    Monthly = 'monthly'
+}
+
+export enum Answered {
+    Yes = 'yes',
+    No = 'no',
+    Pending = 'pending'
+}
```
Comment (petrovdimitar92) [inline src/entities/meeting.entity.ts]: export both enums in another file

Comment (petrovdimitar92) [reply src/entities/meeting.entity.ts]: enum convention is to have only capital keys (ex. DAILY = 'daily')

File: src/entities/meeting.entity.ts [hunk 4]
```diff
@@ -1,19 +1,37 @@
...
 
 export class MeetingEntity {
     @AutoMap()
-    _id!: Types.ObjectId
+        _id!: Types.ObjectId;
+
+    creator!: Creator;
```
Comment (petrovdimitar92) [inline src/entities/meeting.entity.ts:11]: don't use dto in entities and entities in dtos

File: src/entities/meeting.entity.ts [hunk 5]
```diff
@@ -1,19 +1,37 @@
...
 
     @AutoMap()
-    startTime!: Date;
+        startTime!: Date;
 
     @AutoMap()
-    endTime!: Date;
+        endTime!: Date;
+
+    participants?: Participant[];
```
Comment (petrovdimitar92) [inline src/entities/meeting.entity.ts:22]: move as well

File: src/entities/user.entity.ts [hunk 6]
```diff
@@ -1,13 +1,39 @@
-import {Types} from "mongoose";
-import {AutoMap} from "@automapper/classes";
+import { Types } from 'mongoose';
+import { AutoMap } from '@automapper/classes';
+import { type Creator, MeetingDto } from '../dtos/meeting.dto';
+import { Answered } from './meeting.entity';
+
+export class UserMeeting implements Pick<Creator, 'answered'> {
+    meeting_id!: string;
```
Comment (petrovdimitar92) [inline src/entities/user.entity.ts]: use camelCase

File: src/entities/user.entity.ts [hunk 7]
```diff
@@ -1,13 +1,39 @@
...
+    answered: Answered = Answered.Pending;
+}
+
+export class UserMeetingFull extends MeetingDto {
+    answered!: Answered;
+}
```
Comment (petrovdimitar92) [inline src/entities/user.entity.ts]: move this as well

File: src/router/user.router.ts [hunk 8]
```diff
@@ -1,14 +1,20 @@
...
 export const router = Router();
 
-// router.get('/', (req, res) => userController.getAll(req, res));
-// router.post('/', (req, res) => userController.create(req, res));
-// router.get('/:id', (req, res) => userController.getById(req, res));
-// router.put('/:id', (req, res) => userController.updateById(req, res));
-// router.delete('/:id', (req, res) => userController.deleteById(req, res));
+router.post('/register', isGuest(), async (req, res, next) => { await userController.register(req, res, next); }, createToken);
+router.post('/login', isGuest(), async (req, res, next) => { await userController.login(req, res, next); }, createToken);
+router.get('/logout', isLogged(), (req, res) => userController.logout(req, res));
```
Comment (petrovdimitar92) [inline src/router/user.router.ts]: move login, logout, register into another controller and refactor user.router

File: src/router/meeting-room.router.ts [hunk 9]
```diff
@@ -0,0 +1,10 @@
+import { Router } from 'express';
+import { meetingRoomController } from '../controllers/meeting-room.controller';
+
+export const router = Router();
+
+router.get('/', async (req, res, next) => await meetingRoomController.getAll(req, res, next));
+router.post('/', async (req, res, next) => await meetingRoomController.create(req, res, next));
+router.get('/:_id', async (req, res, next) => await meetingRoomController.getById(req, res, next));
```
Comment (petrovdimitar92) [inline src/router/meeting-room.router.ts]: no need for underscore

File: src/middlewares/guards.ts [hunk 10]
```diff
@@ -0,0 +1,93 @@
...
+import createHttpError from 'http-errors';
+import { plainToClass } from 'class-transformer';
+import { validateDto } from '../handlers/validate-request.handler';
+import { type MeetingDto } from '../dtos/meeting.dto';
+
+function userExistsInMeeteing (username: string, meeting: MeetingDto): boolean {
+    if (meeting.creator === username) {
+        return true;
+    }
+    return !!(meeting.participants?.includes(username));
```
Comment (petrovdimitar92) [inline src/middlewares/guards.ts]: `
return meeting.creator === username && !!(meeting.participants?.includes(username));
`

File: src/middlewares/guards.ts [hunk 11]
```diff
@@ -0,0 +1,93 @@
...
+            next(createHttpError.Unauthorized('You are already logged in!'));
+        }
+    };
+}
+
+export function isPathOwner () {
+    return async (req: Request<PathParamUserMeetingDto>, res: Response, next: NextFunction) => {
+        try {
+            // Transform request params to class
+            const pathParams = plainToClass(PathParamUserDto, req.params);
```
Comment (petrovdimitar92) [inline src/middlewares/guards.ts:37]: use plainToInstance

File: src/middlewares/guards.ts [hunk 12]
```diff
@@ -0,0 +1,93 @@
...
+            next(err);
+        }
+    };
+}
+
+export function hasMeeting () {
+    return async (req: Request<PathParamUserMeetingDto>, res: Response, next: NextFunction) => {
+        try {
+            // Transform request params to class
+            const pathParams = plainToClass(PathParamUserMeetingDto, req.params);
```
Comment (petrovdimitar92) [inline src/middlewares/guards.ts:57]: same

File: src/utils/jwt.util.ts [hunk 13]
```diff
@@ -1,17 +1,18 @@
...
+    const token = jwt.sign({ obj }, TOKEN_SECRET, { expiresIn: '1 day' });
+    res.cookie(COOKIE_NAME, token, { httpOnly: true });
 }
 
-export function verifyToken(token: string, secret: string): JwtPayload {
+export function JWTVerifyToken (token: string, secret: string): JwtPayload {
     return jwt.verify(token, secret) as JwtPayload;
-}
\ No newline at end of file
+}
```
Comment (petrovdimitar92) [inline src/utils/jwt.util.ts:18]: user lower case for functions

File: src/middlewares/auth.middleware.ts [hunk 14]
```diff
@@ -1,32 +1,34 @@
...
 
-export const auth = async (req: Request, res: Response, next: NextFunction) => {
+export const auth = async (req: Request, res: Response, next: NextFunction): Promise<Response | void> => {
     try {
         const token: string = req.cookies[COOKIE_NAME];
         if (!token) {
             req.user = undefined;
             next();
             return;
         }
```
Comment (petrovdimitar92) [inline src/middlewares/auth.middleware.ts]: move !token outside of try/catch block

File: src/dtos/meeting-room.dto.ts [hunk 15]
```diff
@@ -0,0 +1,64 @@
...
+        endAvailableHours!: string;
+
+    @IsInt({ message: 'Meeting room capacity must be an integer number!' })
+    @Min(1, { message: 'Meeting room capacity must be at least 1!' })
+    @IsNotEmpty({ message: 'Meeting room capacity is required!' })
+    @Expose()
+    @AutoMap()
+        capacity!: number;
+}
+export class MeetingRoomUpdateDto {
```
Comment (petrovdimitar92) [inline src/dtos/meeting-room.dto.ts]: implement Partial<MeetingRoomDto>

Comment (petrovdimitar92) [reply src/dtos/meeting-room.dto.ts]: MeetingRoomDto

File: src/dtos/meeting.dto.ts [hunk 16]
```diff
@@ -1,32 +1,113 @@
...
+
+    @AutoMap()
+        answered: Answered = Answered.Pending;
+}
+
+export class Creator implements Pick<UserDto, 'username'> {
+    username!: string;
+
+    answered: Answered = Answered.Yes;
+}
```
Comment (petrovdimitar92) [inline src/dtos/meeting.dto.ts]: move Creator & Participant in another category

File: src/dtos/meeting.dto.ts [hunk 17]
```diff
@@ -1,32 +1,113 @@
...
 
 export class MeetingDto {
+    @IsOptional()
```
Comment (petrovdimitar92) [inline src/dtos/meeting.dto.ts:8]: @IsNotEmpty()

File: src/dtos/meeting.dto.ts [hunk 18]
```diff
@@ -1,32 +1,113 @@
...
-    @IsNotEmpty({message: 'Meeting name is required!'})
+    @IsDateString(undefined, { message: 'Meeting end time must of format: YYYY-MM-DDTHH:MM:SS+TIMZ!' })
+    @IsNotEmpty({ message: 'Meeting end time is required!' })
     @Expose()
     @AutoMap()
-    endTime!: Date;
+        endTime!: Date;
 
-    @IsNotEmpty({message: 'Meeting room is required!'})
+    @IsOptional()
```
Comment (petrovdimitar92) [inline src/dtos/meeting.dto.ts:34]: @isarray()@IsNotEmpty({each: true})

File: src/controllers/meeting-room.controller.ts [hunk 19]
```diff
@@ -0,0 +1,86 @@
...
+            return res.status(201).json(created);
+        } catch (err: unknown) {
+            next(err);
+        }
+    }
+
+    async getById (req: Request<PathParamMeetingRoomDto>, res: Response, next: NextFunction): Promise<Response | void> {
+        try {
+            // Validate request params ID
+            const id: string = req.params._id.trim();
```
Comment (petrovdimitar92) [inline src/controllers/meeting-room.controller.ts]: validate req.params

File: src/controllers/meeting-room.controller.ts [hunk 20]
```diff
@@ -0,0 +1,86 @@
...
+            if (!id) {
+                throw createHttpError.BadRequest('Meeting ID missing!');
+            }
```
Comment (petrovdimitar92) [inline src/controllers/meeting-room.controller.ts]: move this checks into managers/services

File: src/controllers/user.controller.ts [hunk 21]
```diff
@@ -1,60 +1,152 @@
...
+            const updated = await this.userService.update(req.params.username, userDto);
+            return res.status(200).json(updated);
+        } catch (err: unknown) {
+            next(err);
+        }
+    }
+
+    async getAllMeetings (req: Request<{}, {}, {}, ReqQueryFilterMeetings>, res: Response, next: NextFunction): Promise<Response | void> {
+        try {
+            const meetings = await this.userManager.getAllMeetings(req.user!._id, req.query.answered, req.query.period);
```
Comment (petrovdimitar92) [inline src/controllers/user.controller.ts]: no !

File: src/services/user.service.ts [hunk 22]
```diff
@@ -1,51 +1,83 @@
...
+        const updated = await this.userRepository.updateById(id, userDto);
+        if (updated == null) {
```
Comment (petrovdimitar92) [inline src/services/user.service.ts]: if(!updated)

**Comment by alexurumov** [comment]
Refactored all according to reviews

**Pull request merged**